### PR TITLE
Fix borderpatrol example module and weave core and auth module together

### LINKED ISSUE #78: Fix borderpatrol example module
The example module should weave core and auth modules along with the business logic.

### PR DESCRIPTION
**Pull request by maheshkelkar**
Fixes #78

### CODE REVIEW

**Comment by maheshkelkar** [comment]
@trane please note that coverage went up, as I excluded the "example" from the code coverage

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala [hunk 1]
```diff
@@ -28,7 +27,7 @@ object Keymaster {
    * @param path Keymaster service endpoint path
    */
   case class KeymasterIdentityProvider(client: Service[Request, Response], path: Path)
-      extends IdentityProvider[Credential, Tokens] {
+    extends IdentityProvider[Credential, Tokens] {
```
Comment (trane) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: please keep this 4 spaces in from where the `case class` definition is. it's confusing to have `extends` at the same position as `def`s

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala [hunk 2]
```diff
@@ -57,38 +56,37 @@ object Keymaster {
...
    * - saves the Tokens after a successful login
    * - sends the User to their original request location from before they logged in or the default location based on
-   *   their service
+   * their service
    * @param store
    * @param secretStoreApi
    */
   case class KeymasterLoginFilter(store: SessionStore)(implicit secretStoreApi: SecretStoreApi)
-      extends Filter[ServiceRequest, Response, IdentifyRequest[Credential], IdentifyResponse[Tokens]] {
+    extends Filter[SessionIdRequest, Response, IdentifyRequest[Credential], IdentifyResponse[Tokens]] {
```
Comment (trane) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: here, too

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala [hunk 3]
```diff
@@ -57,38 +56,37 @@ object Keymaster {
...
 
-    def createIdentifyReq(req: ServiceRequest): Option[IdentifyRequest[Credential]] =
+    def createIdentifyReq(req: SessionIdRequest): Option[IdentifyRequest[Credential]] =
       for {
-        u <- req.req.params.get("username")
-        p <- req.req.params.get("password")
-      } yield KeymasterIdentifyReq(Credential(u, p, req.id))
+        u <- req.req.req.params.get("username")
```
Comment (trane) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: wow, that `req.req.req` would go away if i'd modeled this better. sorry.

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala [hunk 4]
```diff
@@ -103,7 +101,7 @@ object Keymaster {
    * @param store Session store
    */
   case class KeymasterAccessIssuer(client: Service[Request, Response], path: Path, store: SessionStore)
-      extends AccessIssuer[Tokens, ServiceToken] {
+    extends AccessIssuer[Tokens, ServiceToken] {
```
Comment (trane) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: here, too

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala [hunk 5]
```diff
@@ -135,4 +133,17 @@ object Keymaster {
       )(t => Future.value(t)).map(t => KeymasterAccessRes(Access(t)))
   }
 
+  /**
+   * This filter acquires the access and then forwards the request to upstream service
+   *
+   * @param upstreamClient
+   */
+  case class KeymasterAccessFilter(upstreamClient: Service[Request, Response])
+    extends Filter[AccessIdRequest[Tokens], Response, AccessRequest[Tokens], AccessResponse[ServiceToken]] {
```
Comment (trane) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: here, too

File: auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala [hunk 6]
```diff
@@ -135,4 +133,17 @@ object Keymaster {
...
+    def apply(req: AccessIdRequest[Tokens],
+              accessService: Service[AccessRequest[Tokens], AccessResponse[ServiceToken]]): Future[Response] =
+      accessService(AccessRequest(req.id, req.req.req.serviceId, req.req.sid)).flatMap(accessResp =>
```
Comment (trane) [inline auth/src/main/scala/com/lookout/borderpatrol/auth/keymaster/Keymaster.scala]: we should definitely clean this `req.req.req.req` and `access.access` stuff up soon. not for this pr.

File: core/src/main/scala/com/lookout/borderpatrol/sessionx/SessionStore.scala [hunk 7]
```diff
@@ -90,9 +90,11 @@ object SessionStores {
         case _ => None.toFuture
       }
 
-    def update[A](session: Session[A])(implicit ev: SessionDataEncoder[A]): Future[Unit] =
+    def update[A](session: Session[A])(implicit ev: SessionDataEncoder[A]): Future[Unit] = {
+      delete(session.id)
```
Comment (trane) [inline core/src/main/scala/com/lookout/borderpatrol/sessionx/SessionStore.scala]: interesting, why do we need to `delete`? i bet if we set line L95 to `store.set`

Comment (maheshkelkar) [inline core/src/main/scala/com/lookout/borderpatrol/sessionx/SessionStore.scala]: For InMemoryStore, I don't see any "set" method or a overloaded addOrUpdate method.

**Comment by maheshkelkar** [comment]
@trane I have taken care of all comments, except delete/store.set

**Pull request merged**